File: CountGoodNodesinBinaryTree/count_good_nodes_in_binary_tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Solution {
public:
    int goodNodes(TreeNode* root) {
        return dfs(root, root->val);
    }
private:
    int dfs(TreeNode* node, int branch_max){
        if ( !node ) return 0;
        
        branch_max = std::max(branch_max, node->val);
        int count {(node->val == branch_max) ? 1 : 0};
        count += dfs(node->left, branch_max);
        count += dfs(node->right, branch_max);
        return count;
    }
};

File: CountGoodNodesinBinaryTree/count_good_nodes_in_binary_tree.py
#!/usr/bin/env python3

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        def helper(node, branch_max):
            if not node:
                return 0
                        
            branch_max = max(branch_max, node.val)
            count = 1 if node.val == branch_max else 0
            count += helper(node.left, branch_max)
            count += helper(node.right, branch_max)
            return count

        return helper(root, root.val)

